py_trees/behaviour.py

```diff
@@ -75,21 +79,34 @@ def __init__(self, name="", *args, **kwargs):
 
     def setup(self, timeout):
         """
-        Subclasses may override this method to do any one-time delayed construction that
-        is necessary for runtime. This is best done here rather than in the constructor
-        so that trees can be instantiated on the fly without any severe runtime requirements
-        (e.g. a hardware sensor) on any pc to produce visualisations such as dot graphs.
-
+        Subclasses may override this method to do any one-off delayed construction &
+        validation that is necessary prior to ticking the tree. Such construction is best
+        done here rather than in __init__ so that trees can be instantiated on the fly for
+        easy rendering to dot graphs without imposing runtime requirements (e.g. establishing
+        a middleware connection to a sensor).
+        
+        Equally as important, executing methods here which validate the configuration of
+        behaviours will help increase confidence that your tree will successfully tick
+        without logical software errors before actually ticking. This is useful both
+        before a tree's first tick and immediately after any modifications to a tree
+        has been made between ticks.
+        
+        .. tip::
+
+           * If there are children, be sure to recursively call this function on each one so
+             that a single setup invocation at the root of a tree will traverse the entire tree.
+           * Faults are notified to the user of the behaviour via exceptions. Choice of exception to
+             use is left to the user.
+        
         .. note:: User Customisable Callback
 
         Args:
-            timeout (:obj:`float`): time to wait (use common.Duration.INFINITE to block indefinitely)
+            timeout (:obj:`float`): time (s) to wait (use common.Duration.INFINITE to block indefinitely)
 
-        Returns:
-            :obj:`bool`: whether it timed out trying to setup
+        Raises:
+            Exception: if this behaviour has a fault in construction or configuration
         """
         del timeout  # Unused argument
```

Comment:
BTW I was just thinking, does it make sense for this to take a timeout parameter. Now that we have exceptions (and the parameter is unused), we should let the user throw one if something takes too long to load. Thoughts?

Reply:
Ah, I'd forgotten the clunkiness of this. The timeout really needs to be dictated by the owner of the tree (especially for robots) but getting the owner of the behaviour dealing with it is, really, just random noise huh.

In addition, that timeout is actually N*timeout, where N is the number of behaviours in the tree, when handled this way (hence clunky). Hence it's not even a good mechanism for the owner of the tree.

What about a setup timeout handled in trees.BehaviourTree? I think that makes more sense. Will follow up in a different PR.

Reply:
> What about a setup timeout handled in trees.BehaviourTree?

:+1: